opencog/atoms/execution/CondLink.cc

```diff
@@ -0,0 +1,99 @@
+ * it under the terms of the GNU Affero General Public License v3 as
+ * published by the Free Software Foundation and including the
+ * exceptions at http://opencog.org/wiki/Licenses
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public
+ * License along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <opencog/atoms/base/ClassServer.h>
+#include <opencog/atoms/execution/Instantiator.h>
+#include <opencog/atoms/core/FindUtils.h>
+
+#include "CondLink.h"
+#include "EvaluationLink.h"
+
+using namespace opencog;
+
+void CondLink::init(void)
+{
+	if (0 == _outgoing.size())
+		throw SyntaxException(TRACE_INFO,
+		                      "CondLink is expected to be arity greater-than 0!");
+
+	if (1 == _outgoing.size()) {
+		default_exp = _outgoing[0];
+		return;
+	}
+
+	for (unsigned i = 0; i < _outgoing.size(); ++i) {
+		// If the conditions and expressions are wrapped in list_link
+		if (LIST_LINK == _outgoing[i]->get_type()) {
+			// The first item in the list_link holds the condition,
+			// and the second holds the expression.
+			conds.push_back(_outgoing[i]->getOutgoingSet()[0]);
+			exps.push_back(_outgoing[i]->getOutgoingSet()[1]);
+			continue;
+		}
+
+		// If cond_link starts wrapping conds and exps in list_link, then it is
+		// expected to be consistent. If one wants to have a default expression
+		// using true_link as condition should do it.
+		if (i != 0 && LIST_LINK == _outgoing[i - 1]->get_type())
+			throw SyntaxException(TRACE_INFO,
+			                      "CondLink is expected to wrap expressions in LIST_LINK.");
+
+		// If the conditions and expressions are flattened in even and odd
+		// positions respectively.
+		if (i % 2 == 0) {
+			if (i == _outgoing.size() - 1) {
+				default_exp = _outgoing[i];
+				return;
+			}
+			conds.push_back(_outgoing[i]);
+		} else {
+			exps.push_back(_outgoing[i]);
+		}
+	}
+}
+
+CondLink::CondLink(const HandleSeq &oset, Type t)
+		: FunctionLink(oset, t)
+{
+	if (not nameserver().isA(t, COND_LINK)) {
+		const std::string &tname = nameserver().getTypeName(t);
+		throw SyntaxException(TRACE_INFO,
+		                      "Expecting a CondLink, got %s", tname.c_str());
+	}
+
+	// Derived types have a different initialization sequence.
+	if (COND_LINK != t) return;
+	init();
+}
+
+Handle CondLink::execute(AtomSpace *scratch) const
+{
+	TruthValuePtr tvp;
+	Instantiator inst(scratch);
+	for (unsigned i = 0; i < conds.size(); ++i) {
+		tvp = EvaluationLink::do_evaluate(scratch, conds[i]);
+		if (tvp->get_mean() > 0.5)
+			return HandleCast(inst.instantiate(exps[i], HandleMap()));
```

Comment:
Hi @linas, In case of eager execution the expressions in the CondLink might have been already executed right? such as an ArithmeticLink might have been already reduced to a NumberNode and calling execute on that NumberNode should throw the "Not Executable" exception in Atom.h? or am I missing something?

Reply:
Yes, for eager evaluation, the arguments will already have been reduced. There should not be any exceptions as a result, as the reduction is (meant to be) performed only once, not twice. 

Lay evaluation would be better; it almost works; there's some open issue saying "must finish this work. Anyway, toe CondLink code looks OK to me.